packages/layout-elk/src/glsp-elk-layout-engine.ts

```diff
@@ -13,45 +12,321 @@
+                }
+            });
+            return graph;
+        } else if (model instanceof GNode) {
+            return this.transformNode(model);
+        } else if (model instanceof GEdge) {
+            return this.transformEdge(model);
+        } else if (model instanceof GLabel) {
+            return this.transformPort(model);
+        } else if (model instanceof GPort) {
+            return this.transformPort(model);
+        }
+
+        throw new Error('Type not supported: ' + model.type);
+    }
+
+    /**
+     * Searches for all children of the given element that are an instance of the given {@link GModelElementConstructor}
+     * and are included by the {@link ElementFilter}. Also considers children that are nested inside of {@link GCompartment}s.
+     * @param element The element whose children should be queried.
+     * @param constructor The class instance that should be matched
+     * @returns A list of all matching children.
+     */
+    protected findChildren<G extends GModelElement>(element: GModelElement, constructor: GModelElementConstructor<G>): G[] {
+        const result: G[] = [];
+        element.children.forEach(child => {
+            if (child instanceof constructor && this.filter.apply(element)) {
+                result.push(child);
+            } else if (child instanceof GCompartment) {
+                result.push(...this.findChildren(child, constructor));
+            }
+        });
+
+        return result;
+    }
+
+    protected findCommonAncestor(elkEdge: ElkPrimitiveEdge): ElkNode | undefined {
+        const source = this.modelState.index.get(elkEdge.source);
+        const target = this.modelState.index.get(elkEdge.target);
+        if (!source || !target) {
+            return undefined;
+        }
+
+        const sourceParent = findParent(source.parent, parent => parent instanceof GNode || parent instanceof GGraph);
+        const targetParent = findParent(target.parent, parent => parent instanceof GNode || parent instanceof GGraph);
+
+        if (!sourceParent || !targetParent) {
+            return undefined;
+        }
+
+        if (sourceParent === targetParent) {
+            return this.idToElkElement.get(sourceParent.id);
+        } else if (source === targetParent) {
+            return this.idToElkElement.get(source.id);
+        } else if (target === sourceParent) {
+            return this.idToElkElement.get(target.id);
+        }
+        return undefined;
+    }
+
+    protected transformGraph(graph: GGraph): ElkGraphElement {
+        const elkGraph: ElkNode = {
+            id: graph.id,
+            layoutOptions: this.configurator.apply(graph)
+        };
+        if (graph.children) {
+            elkGraph.children = this.findChildren(graph, GNode).map(child => this.transformToElk(child)) as ElkNode[];
+            elkGraph.edges = [];
+            this.elkEdges.push(...(this.findChildren(graph, GEdge).map(child => this.transformToElk(child)) as ElkEdge[]));
```

Comment:
Should these edges not also be pushed into the `elkGraph.edges`? Otherwise the array is always empty, right?

Reply:
No, edges need special handling. ELK expects that edges are always set in their respective parent container (either the graph or another node in case of nested diagrams). Otherwise edges want be layouted correctly. However, the same assumption is not true for the graphical model. If we have an edge between A and B the edge doesn't have to be a child of the container of A and B. It could be placed anywhere in the graph (e.g. as child of A).

Hence I implemented the same mechanism that is used in the Java LayoutEngine. (Store the edges in a temporary map and then in a postprocess step find the most suitable parent container). (https://github.com/eclipse-glsp/glsp-server-node/blob/54b5138f7867d0a73862e5bbac0d350deca8ec3f/packages/layout-elk/src/glsp-elk-layout-engine.ts#L76)